Python/immutability.c

```diff
@@ -0,0 +1,376 @@
+    if(frozen_builtins == NULL){
+        goto nomemory;
+    }
+
+    frozen_globals = PyDict_New();
+    if(frozen_globals == NULL){
+        goto nomemory;
+    }
+
+    while(PyList_Size(f_stack) != 0){
+        f_ptr = pop(f_stack); // fp.rc = x + 1
+        _PyObject_ASSERT(f_ptr, PyCode_Check(f_ptr));
+        f_code = (PyCodeObject*)f_ptr;
+
+        size = 0;
+        if (f_code->co_names != NULL)
+          size = PySequence_Fast_GET_SIZE(f_code->co_names);
+        for(Py_ssize_t i = 0; i < size; i++){
+            PyObject* name = PySequence_Fast_GET_ITEM(f_code->co_names, i); // name.rc = x
+
+            if(PyUnicode_CompareWithASCIIString(name, "globals") == 0){
+                // if the code calls the globals() builtin, then any
+                // cellvar or const in the function could, potentially, refer to
+                // a global variable. As such, we need to check if the globals
+                // dictionary contains that key and then make it immutable
+                // from this point forwards.
+                check_globals = true;
+            }
+
+            if(PyDict_Contains(globals, name)){
+                PyObject* value = PyDict_GetItem(globals, name);
+                if(PyDict_SetItem(frozen_globals, name, value)){
+                    Py_DECREF(frozen_builtins);
+                    Py_DECREF(frozen_globals);
+                    Py_DECREF(f_stack);
+                    return NULL;
+                }
+            }else if(PyDict_Contains(builtins, name)){
+                PyObject* value = PyDict_GetItem(builtins, name);
+                if(PyDict_SetItem(frozen_builtins, name, value)){
+                    Py_DECREF(frozen_builtins);
+                    Py_DECREF(frozen_globals);
+                    Py_DECREF(f_stack);
+                    return NULL;
+                }
+            }else if(PyDict_Contains(module_dict, name)){
+                PyObject* value = PyDict_GetItem(module_dict, name); // value.rc = x
+
+                _PyDict_SetKeyImmutable((PyDictObject*)module_dict, name);
+
+                if(!_Py_IsImmutable(value)){
+                    if(PyList_Append(frontier, value)){
+                        goto nomemory;
+                    }
+                }
+            }
+        }
+
+        size = PySequence_Fast_GET_SIZE(f_code->co_consts);
+        for(Py_ssize_t i = 0; i < size; i++){
+            PyObject* value = PySequence_Fast_GET_ITEM(f_code->co_consts, i); // value.rc = x
+            if(!_Py_IsImmutable(value)){
+                if(PyCode_Check(value)){
+                    _Py_SetImmutable(value);
+
+                    if(PyList_Append(f_stack, value)){
+                        goto nomemory;
+                    }
+                }else{
+                    if(PyList_Append(frontier, value)){
+                        goto nomemory;
+                    }
+                }
+            }
+
+            if(check_globals && PyUnicode_Check(value)){
```

Comment:
This `if`-block was a bit hard to understand. Is my current understanding correct:

When a function accesses `globals()` it might use a string to get a value out of global. For example, the `"hej"` in this example `globals()["hej"]`. This check freezes and transfers the value of `"hej"` into the `frozen_globals`.

If this understanding is correct, could you maybe add a small comment?

Reply:
yup, that's right. I'll add a comment.